src/plugins/clips-executive/clips/domain.clp

```diff
@@ -178,6 +178,260 @@
+  has not been grounded yet and add a reference to the plan-action fact"
+  (declare (salience ?*SALIENCE-DOMAIN-GROUND*))
+  ?p <- (plan-action (id ?action-id) (action-name ?operator-id) 
+                     (param-names $?param-names) (param-values $?param-values) 
+                     (preconditions $?precons&:(= (length$ ?precons) 0)))
+  (domain-operator (name ?operator-id) (param-names $?op-param-names&:(= (length$ ?param-names) (length$ ?op-param-names))))
+  =>
+  (bind ?grounded-formulas (ground-pddl-formula ?operator-id "" ?param-names ?param-values ""))
+  (modify ?p (preconditions ?grounded-formulas))
+)
+
+(defrule domain-check-if-action-precondition-is-satisfied
+  "Check if there is a referenced precondition formula that is satisfied,
+  if yes make the action executable."
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  ?p <- (plan-action (executable FALSE) (id ?id) (preconditions $? ?formula-id $?))
+  (grounded-pddl-formula (is-satisfied TRUE) (id ?formula-id))
+  =>
+  (modify ?p (executable TRUE))
+  (printout t crlf crlf crlf "Action " ?id " is executable based on " ?formula-id crlf crlf crlf)
+)
+
+(defrule domain-check-if-action-precondition-is-unsatisfied
+  "Check if all referenced precondition formulas are not satisfied,
+  if yes make the action not executable."
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  ?p <- (plan-action (executable TRUE) (id ?id) (preconditions $? ?formula-id $?))
+  (not (grounded-pddl-formula (is-satisfied TRUE) (id ?formula-id)))
+  =>
+  (modify ?p (executable FALSE))
+  (printout t crlf crlf crlf "Action " ?id " is no longer executable" crlf crlf crlf)
+)
+
+(defrule domain-check-if-negated-formula-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied FALSE))
+  (pddl-formula (id ?base) (type negation))
+
+  (or (grounded-pddl-formula (is-satisfied FALSE) (part-of ?id) (id ?child-id&:(neq ?id ?child-id)))
+      (grounded-pddl-predicate (is-satisfied FALSE) (part-of ?id))
+  )
+
+=>
+  (modify ?parent (is-satisfied TRUE))
+)
+
+(defrule domain-check-if-negated-formula-is-unsatisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied TRUE))
+  (pddl-formula (id ?base) (type negation))
+
+  (or (grounded-pddl-formula (is-satisfied TRUE) (part-of ?id) (id ?child-id&:(neq ?id ?child-id)))
+      (grounded-pddl-predicate (is-satisfied TRUE) (part-of ?id))
+  )
+=>
+  (modify ?parent (is-satisfied FALSE))
+)
+
+(defrule domain-check-if-conjunctive-formula-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied FALSE))
+  (pddl-formula (id ?base) (type conjunction))
+
+  (not (grounded-pddl-formula (part-of ?id) (is-satisfied FALSE) (id ?child-id&:(neq ?id ?child-id))))
```

Comment:
This would be unnecessary if the formula would never have itself as parent (see above)